python/neuroglancer/pipeline/storage.py

```diff
@@ -190,15 +227,106 @@ def _uncompress(content):
             return gfile.read()
 
     def list_files(self, prefix=""):
-        for f in self._interface(self._path).list_files(prefix):
+        for f in self._interface.list_files(prefix):
             yield f
 
-    def wait_until_queue_empty(self):
-        if self._n_threads:
+    def wait(self):
+        if len(self._threads):
             self._queue.join()
+        return self
+
+    def __del__(self):
+        self.kill_threads()
+        self._interface.release_connection()
+
+    def __enter__(self):
+        self.start_threads(self._n_threads)
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        self.wait()
+        self.kill_threads()
+        self._interface.release_connection()
+
+class ConnectionPool(object):
```

Comment:
I would explain what's the intent of a connection pool or responsabilities.
Why isn't it part of the Interfaces